src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -197,51 +177,108 @@ private void EmitParametersIfPresent(ExpressionEmitter emitter, ImmutableArray<D
 
-                    var functionCall = ExpressionBuilder.Convert(decoratorSyntax.Expression) as FunctionCallExpression
-                        ?? throw new InvalidOperationException($"Failed to convert decorator expression {decoratorSyntax.Expression.GetType()}");
+            foreach (var (modifier, propertyName) in new[]
+            {
+                (expression.Metadata, LanguageConstants.ParameterMetadataPropertyName),
+                (expression.MinLength, LanguageConstants.ParameterMinLengthPropertyName),
+                (expression.MaxLength, LanguageConstants.ParameterMaxLengthPropertyName),
+                (expression.MinValue, LanguageConstants.ParameterMinValuePropertyName),
+                (expression.MaxValue, LanguageConstants.ParameterMaxValuePropertyName),
+            }) {
+                if (modifier is not null)
+                {
+                    modifications.Add((modifier, result => result.MergeProperty(propertyName, modifier)));
+                }
+            }
 
-                    var evaluated = decorator.Evaluate(functionCall, targetType, result);
-                    if (evaluated is not null)
+            if (expression.Description is {} description)
+            {
+                modifications.Add((description, result => ApplyDescription(expression, result)));
+            }
+
+            // Whether one decorator overrides another is determined by their order in the syntax tree
+            // TODO should we change this?
+            if (expression.SourceSyntax is DecorableSyntax decorable)
+            {
+                ConcurrentDictionary<(Expression, Func<ObjectExpression, ObjectExpression>), int> modificationIndices = new();
+                int GetIndex((Expression, Func<ObjectExpression, ObjectExpression>) modification)
+                {
+                    if (modification.Item1.SourceSyntax is {} sourceSyntax)
                     {
-                        result = evaluated;
+                        int idx = 0;
+                        foreach (var decorator in decorable.Decorators)
+                        {
+                            if (Context.SemanticModel.Binder.IsDescendant(sourceSyntax, decorator))
+                            {
+                                return idx;
+                            }
+
+                            idx++;
+                        }
                     }
+
+                    return int.MaxValue;
                 }
+
+                modifications.Sort((a, b) => modificationIndices.GetOrAdd(b, GetIndex) - modificationIndices.GetOrAdd(a, GetIndex));
+            }
+
+            var result = input;
+            foreach (var modification in modifications)
+            {
+                result = modification.modifyFunc(result);
             }
 
             return result;
         }
 
+        private ObjectExpression ApplyDescription(DescribableExpression expression, ObjectExpression input) => expression.Description is {} description
```

Comment:
It seems the method can be marked as static.

Reply:
I went through and marked as much of the type expression handling `static` as I could. The only reason any of the methods need to access to class state is to apply decorators in the order in which they appeared in the Bicep syntax tree. I'll bring up a discussion topic on Wednesday to see if this is something we need to keep doing.

Reply:
Opened #11308 to track.